src/main/java/orlov/programming/springcoregym/dao/impl/user/trainee/TraineeDaoImpl.java

```diff
@@ -1,49 +1,71 @@
-
-    private static final String TRAINEE_NULL_ERROR = "Trainee can't be null";
-
-    @Autowired
-    public TraineeDaoImpl(Storage storage) {
-        super(storage, Trainee.class);
-    }
+public class TraineeDaoImpl extends AbstractDao<Trainee, Long> implements TraineeDao {
 
     @Override
-    protected String getNullEntityErrorMessage() {
-        return TRAINEE_NULL_ERROR;
+    protected Class<Trainee> getEntityClass() {
+        return Trainee.class;
     }
 
     @Override
-    protected void checkEntityIdIsNull(Long id) {
-        if (id == null) {
-            IllegalArgumentException e = new IllegalArgumentException("Trainee's id can't be null");
-            log.error(e);
-            throw e;
+    public Optional<Trainee> findByUsername(String username) {
+        try {
+            String jpql = "SELECT t FROM " + getEntityClass().getSimpleName() + " t WHERE t.username = :username";
+            TypedQuery<Trainee> query = getEntityManager().createQuery(jpql, Trainee.class);
+            query.setParameter("username", username);
+
+            Trainee trainee = query.getSingleResult();
+            return Optional.of(trainee);
+
+        } catch (NoResultException e) {
+            log.info("No trainee found with username: {}", username);
+            return Optional.empty();
         }
     }
 
     @Override
-    protected Long getUserId(Trainee trainee) {
-        return trainee.getUserId();
-    }
+    public List<Training> getTrainingsByDateUsernameTrainingType
+            (LocalDate startDate, LocalDate endDate, String userName, String trainingType) {
+        String jpql = "SELECT tr FROM Training tr "
+                + "JOIN tr.trainee t "
+                + "JOIN tr.trainingType tt "
+                + "WHERE t.username = :username "
+                + "AND tt.trainingTypeName = :trainingType "
+                + "AND tr.trainingDate BETWEEN :startDate AND :endDate";
 
-    @Override
-    protected void setUserId(Trainee trainee, Long id) {
-        trainee.setUserId(id);
+        TypedQuery<Training> query = getEntityManager().createQuery(jpql, Training.class);
+        query.setParameter("username", userName);
+        query.setParameter("trainingType", trainingType);
+        query.setParameter("startDate", startDate);
+        query.setParameter("endDate", endDate);
+
+        return query.getResultList();
     }
 
+    @Transactional
     @Override
-    protected String getUsername(Trainee trainee) {
-        return trainee.getUsername();
+    public void deleteByUsername(String username) {
+        String jpql = "DELETE FROM " + getEntityClass().getSimpleName() + " t WHERE t.username = :username";
+        Query query = getEntityManager().createQuery(jpql);
+        query.setParameter("username", username);
+        int rowsAffected = query.executeUpdate();
+
+        log.info("Deleted {} trainee(s) with username: {}", rowsAffected, username);
```

Comment:
do we need it?
maybe it is better to make such a method to return a boolean value (true - deleted, false - was not deleted)